grpc/transport-auth.go

```diff
@@ -0,0 +1,87 @@
+package grpc
+
+import (
+	"crypto/tls"
+	"crypto/x509"
+	"errors"
+	"fmt"
+	"net"
+	"time"
+
+	"golang.org/x/net/context"
+	"google.golang.org/grpc/credentials"
+)
+
+// transportCredentials is a grpc/credentials.TransportCredentials which supports
+// connecting to, and verifying multiple DNS names
+type transportCredentials struct {
+	configs map[string]*tls.Config
+}
+
+func newTransportCredentials(addrs []string, rootCAs *x509.CertPool, clientCerts []tls.Certificate) (credentials.TransportCredentials, error) {
+	configs := make(map[string]*tls.Config, len(addrs))
+	for _, addr := range addrs {
+		host, _, err := net.SplitHostPort(addr)
+		if err != nil {
+			return nil, err
+		}
+		configs[addr] = &tls.Config{
+			ServerName:   host,
+			RootCAs:      rootCAs,
+			Certificates: clientCerts,
+		}
+	}
+	return &transportCredentials{configs}, nil
+}
+
+// ClientHandshake performs the TLS handshake for a client -> server connection
+func (tc *transportCredentials) ClientHandshake(addr string, rawConn net.Conn, timeout time.Duration) (net.Conn, credentials.AuthInfo, error) {
+	ctx := context.Background()
```

Comment:
Also, did you look into this question? If the answer is no we should have a comment about why:

> And maybe we can use conn.SetDeadline(..) instead of doing our own timeout?